### PR TITLE
Mobile: one module for build/OTA state, and a native QR that lands on the PR

### PR DESCRIPTION
**Pull request by mmkal**
Scanning a PR's native-build QR should put you on that PR — the right binary
*and* the right OTA channel. Today it puts you on main, and the link that would
fix that is folded away in a collapsed `<details>` exactly when you need it.
This makes the QR mean what it says, and pulls the scattered "which build am I
on?" state into one module so the screens can be dumb.

## Why the native QR lands you on main

Four separate causes, all live:

1. The install QR is an `expo.dev` build page URL. It carries no channel, and a
   `preview`-profile binary boots on channel `preview` — which is main.
2. The two QRs are mutually-exclusive `<details open>` blocks. When the runtime
   differs — the one case where you need *both*, install then switch — the
   switch link is the collapsed one.
3. `resetChannelOverrideForNewInstall()` force-clears the override on a new
   binary's first boot. Correct in general, but it means install-then-switch
   has a strict order, and nothing says so.
4. The heuristic picking which QR to expand compares against the newest
   finished EAS build, not against what is on your phone.

### Per-PR native builds

A PR's build is made *for that PR's channel*, so installing it is being on the
PR's JS — no second scan. The section says so: **"This build boots on
`<channel>` — installing it is all you need."**

`ensureBuildForPr({channel, runtime})` matches on channel as well as runtime,
prefers a **finished** build, and renders "build still running" instead of
linking a queued one as though it were installable. One build per PR *branch*,
reused across pushes.

Both QRs are now individually correct, so guessing wrong costs a scan rather
than an afternoon — and the new-install guard stops being a footgun, because
clearing the override now lands you where you wanted.

### The fingerprint: the plan was wrong, and measuring it was the point

The plan claimed the channel was safe in `eas.json` because fingerprint never
reads it. It does:

```
{"type":"file","filePath":"eas.json","reasons":["easBuild"],"hash":"331eee28…"}
```

Measured: changing one profile's `channel` moved the fingerprint **fdfdff89 →
793605ca**. Rewriting `eas.json` per PR would have given each PR's build its own
runtime version — a binary that refuses the very updates its PR publishes. The
exact failure the plan thought it was avoiding. (`preview` and `production`
share a runtime today only because they share a *file*; the bytes don't differ
per profile.)

So `fingerprint.config.js` now sets `ignorePaths: ["eas.json"]`. Re-measured
with the ignore in place: **fca56340 either side** of the channel change.

Two consequences, both accepted and documented in the config:

- The baseline hash moves once, so installed binaries need one rebuild. CI
  already triggers one when nothing matches.
- `eas.json`'s other fields (`distribution`, `developmentClient`, `simulator`)
  stop bumping the runtime. They shape which *binary* you get, not which JS a
  binary can run, and each already lives on its own channel.

### One module owns the state

`lib/build-state-core.ts` is the pure owner of every rule — which channel am I
really on, is this build worth watching, what does an update check mean — with
11 unit tests in the node lane. `lib/build-state.ts` is its Expo/react-query
binding and absorbs `build-info.ts`, `preview-channel.ts` and
`native-install-guard.ts`, which are gone.

```ts
const state = useBuildState();
// state.channel               → "mobile-build-state"   (where JS comes from)
// state.binary.defaultChannel → "mobile-build-state"   (what this build is for)
// state.running.message       → "Mobile: build a PR's binary for its own channel"
// state.update                → { kind: "behind", message: "fix the drawer glyph", … }
```

Build info and the QR confirm screen are now views over that: ten-odd queries
and mutations between them became `useBuildState()` + `useBuildActions()`.

Build info

### Freshness on open

On a watched build — an overridden channel, or a binary built for something
other than main — a check on mount and on every foreground, surfaced as a
banner with **Update now** rather than a reload under your fingers. The
foreground trigger is react-query's `focusManager`, which `query.ts` already
wires to `AppState`, so this needs no listener of its own. A phone tracking
main stays exactly as quiet as before.

### App-global sign-in

`lib/session.ts` is the one answer to "am I signed in, where, and as whom?".
The QR screen's mismatch card used to render a bare **"not signed in"** — which
reads as a claim about your session, when it was really about a deployment you
had never signed in to. It now names the server, and shows your real session
underneath.

QR confirm

A dead sign-in is also handled once, in the query cache's `onError`, instead of
being caught separately in `projects.tsx` and `chat.tsx`.

## Two things the app cannot honestly say

- **"Your binary is too old for this channel."** `checkForUpdateAsync` answers
  `noUpdateAvailableOnServer` both when you're current and when the channel has
  newer JS your runtime can't run — the server filters by runtime *before*
  replying. The copy is "You're on the latest update this build can run" and
  claims nothing more.
- **"You're N commits behind."** Same reason: only "something newer exists" is
  knowable, so the banner names the commit instead of counting.

## Re-audit (second pass, stronger model)

The plan and first two commits ran on a weaker model; this branch was then
re-audited from scratch against the original request. Green CI had hidden five
defects and one missed requirement — all fixed in `654d56689`:

1. **"Check for update" was a silent no-op on main phones** — the common case.
   `refetchQueries` skips disabled queries, and the check query is disabled on
   every unwatched build; the state mapping would have discarded the result
   anyway. Now an imperative `fetchQuery` into the same cache.
2. **No way back to main on a per-PR binary.** Per-PR builds change what
   "Reset to default channel" means — the default *is* the PR channel, and
   cleanup deletes it on merge, stranding the phone. Build info now has
   **Switch to main (preview)** on non-main binaries; the PR-body footer stops
   promising the old route.
3. **Stale update verdict across channel switches** — the check cache
   described the old channel; removed on switch.
4. **Override-read race on the QR screen** — it could act (including the
   freshness *reload*) on the binary's default channel before the stored
   override was read. The hook now exposes `ready`; the screen gates on it.
5. **Answerless taps** — "up-to-date" and "no-update" outcomes now say so.
6. **Missed requirement:** *"after I install it it should show me a screen"* —
   first boot of a new binary now opens Build info (inert on a brand-new
   user's first ever run).

## Risk map

**Riskiest: the fingerprint ignore.** `ignorePaths: ["eas.json"]` is what makes
per-PR channels possible, and it narrows what bumps the runtime. Measured both
directions (above), but the first PR to run this is the first real test of
channel-baking end to end. If it's wrong, the symptom is loud and immediate: a
PR's install QR gives a binary that ignores the PR's updates.

**Second: no device pass.** The web lane can't exercise OTA at all
(`Updates.isEnabled` is false there), so the banner, the on-open check and
install-then-run are unproven on hardware. Everything the web lane *can* reach
is covered — the reworded mismatch card, the sign-in flows, Build info's new
rows — and the rules behind the rest have node tests.

**On merge:** the first push to a mobile PR now waits ~15–20 min for its build
before the install QR works; later pushes reuse it. Worth watching EAS build
minutes. One forced rebuild for installed binaries from the fingerprint move.
No data migration, no stale clients.

**Review order:** `apps/mobile/src/lib/build-state-core.ts` + its test first
(every rule lives there), then `apps/mobile/src/lib/build-state.ts` (the
re-audit bugs all lived in this binding layer — cache interactions deserve the
closest read), then `apps/mobile/fingerprint.config.js` and
`scripts/ci/mobile-preview.ts` (the fingerprint argument and the build
selection), then `src/lib/session.ts`, then the screens.
module is in.

## Status

Implemented and field-tested on a real phone twice — both rounds found real
defects that no test lane could reach (an embedded-bundle first-boot, the
`Updates.channel` lie, invisible tap outcomes), all fixed here. CI green.
The task file, including the two deferred items, is
[`tasks/complete/2026-08-29-mobile-build-state-module.md`](tasks/complete/2026-08-29-mobile-build-state-module.md).

An explainer of the whole system ships with this PR:
[`explainers/mobile-builds-and-ota.html`](explainers/mobile-builds-and-ota.html)
— also live at https://iterate.iterate.app/explainers/mobile-builds-and-ota.

## Not done

Stamping the available update's commit message into app config `extra` so the
banner can say *what* the newer update is. The reader (`stampFromManifest`) is
in and degrades cleanly to "a newer update is available"; the config change
(`app.json` → `app.config.js`, plus `ExpoConfigExtraSection` in `sourceSkips`)
is not, and needs its own fingerprint measurement — the same one this PR just
proved you can't skip.

---

<sub>Claude Code session `726a54ca-dad8-443b-8e01-bfd6ee75ae1f` — "mobile build state module"</sub>

| Group | Lines | Significant |
| --- | ---: | ---: |
| Mobile | +1,468 -461 | +920 -336 🟩🟩🟩🟩🟥 |
| Tests | +109 -5 | +93 -2 🟩⬜⬜⬜⬜ |
| CI & scripts | +113 -28 | +56 -18 🟩⬜⬜⬜⬜ |
| Docs | +657 -376 | +533 -298 🟩🟩🟥⬜⬜ |
| Other | +738 -0 | +728 -0 🟩🟩🟩⬜⬜ |
| **Total** | **+3,085 -870** | **+2,330 -654** 🟩🟩🟩🟩🟥 |

<sub>Lines counts every changed line; Significant ignores blank lines, JS comments, and TypeScript lines with no runtime output. Between 6f29833 and 3cc859a, bucketed first-match-wins into the groups defined in `scripts/ci/loc-report.ts`.</sub>

## Environment Config Lease

<details>
<summary>No preview slot recorded.</summary>

| app | status | commit | preview | size (gzip) | deploy duration | test duration | retries | cleanup duration | workflow run | updated | summary |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Auth | released | `3cc859a` | [https://auth.iterate-preview-16.com](https://auth.iterate-preview-16.com) | 817.4 KiB | 23.5s | 17.7s |  | 3.5s | [Workflow run](https://github.com/iterate/iterate/actions/runs/549160258699566) | 2026-08-29T19:43:14.851Z | Preview app released. |
| Docs | released | `3cc859a` | [https://docs-preview-16.iterate-dev-preview.workers.dev](https://docs-preview-16.iterate-dev-preview.workers.dev) | 866.2 KiB | 18.4s | 2.2s |  | 3.0s | [Workflow run](https://github.com/iterate/iterate/actions/runs/549160258699566) | 2026-08-29T19:43:14.285Z | Preview app released. |
| Dummy Petshop | released | `3cc859a` | [https://dummy-petshop.iterate-preview-16.com](https://dummy-petshop.iterate-preview-16.com) | 198.3 KiB | 54ms | 5.8s |  | 557ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/549160258699566) | 2026-08-29T19:43:11.867Z | Preview app released. |
| OS | released | `3cc859a` | [https://os.iterate-preview-16.com](https://os.iterate-preview-16.com) | 5.14 MiB (-11.3 KiB vs main) | 49.3s | 227.6s | 3 retried: the source-version pin tells a facet that recycled from one the commit rebuilt (vitest x1) · mobile/chat-photos.spec.ts › photos fill their bubble, and a tall one sits on its own blurred backdrop › mobile (playwright x1) — TimeoutError: locator.fill: Timeout 1ms exceeded. Call log: [2m - waiting for getByLabel('Organization name')[22m [33m If this is a slow operation, update the product code to add a spinner while it's running. This will improve the user experience and buy you more time for this assertion. To ad... · mobile/media.spec.ts › renders, searches, and views seeded media › mobile (playwright x1) — TimeoutError: locator.click: Timeout 1ms exceeded. Call log: [2m - waiting for getByLabel('Full screen media')[22m [33m If this is a slow operation, update the product code to add a spinner while it's running. This will improve the user experience and buy you more time for this assertion. To a... | 14.4s | [Workflow run](https://github.com/iterate/iterate/actions/runs/549160258699566) | 2026-08-29T19:43:25.696Z | Preview app released. |

</details>

## 📱 Mobile preview

Channel `mobile-build-state` · runtime `fca563400` · **native changes** — needs a fresh install

<details><summary>OTA — switch the installed app to this PR's channel (`3cc859a`)</summary>

**[Switch this phone to the PR channel](https://os.iterate.com/m/preview-channel/mobile-build-state)**

<a href="https://os.iterate.com/m/preview-channel/mobile-build-state">QR code</a>

</details>
<details open><summary>Full install — if the app is uninstalled or the runtime differs (`4cdeb36`)</summary>

**[Open the EAS build install page](https://expo.dev/accounts/mishanustom/projects/iterate/builds/4a94d33d-a0c9-456f-842c-1779e8a99c94)**

<a href="https://expo.dev/accounts/mishanustom/projects/iterate/builds/4a94d33d-a0c9-456f-842c-1779e8a99c94">QR code</a>

<sub>This build boots on <code>mobile-build-state</code> — installing it is all you need.</sub>

</details>

<sub>Back to main inside the app: Build info → Switch to main. Republishes on every push to this PR.</sub>

---

> [!NOTE]
> **High Risk**
> Fingerprint ignore and per-PR channel baking affect which binaries accept which OTA updates; mistakes show up as installs that ignore PR updates, and OTA/channel edge cases are only partly covered by web and unit tests.
> 
> **Overview**
> **PR native installs now boot on the PR’s channel** so the install QR is enough—no second OTA scan. CI adds a `preview-pr` EAS profile, rewrites that profile’s `channel` before `eas build`, and `ensureBuildForPr` picks or triggers a build matching both runtime and channel (with “build still running” when it isn’t finished). **`fingerprint.config.js` ignores `eas.json`** so baking channels per PR doesn’t change the runtime fingerprint and break OTA.
> 
> **Build and update logic is centralized** in `build-state-core.ts` (pure rules + tests) and `build-state.ts` (Expo/react-query), replacing `build-info.ts`, `preview-channel.ts`, and `native-install-guard.ts`. **Build info**, the QR confirm screen, and root layout consume `useBuildState()` / `useBuildActions()`: clearer channel vs default rows, revert-on-failed “switch to main”, first-boot navigation to Build info instead of an alert, and an **`UpdateBanner`** on “watched” builds (foreground checks via existing focus wiring).
> 
> **Sign-in is app-global** via new `lib/session.ts`; `query.ts` handles `SignInRequiredError` once. Docs, an interactive explainer HTML, and CI/PR-body copy reflect the new two-QR flow and “Switch to main” wording.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3cc859a67fc201297da65256c43fbfb710ea3c0e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by mmkal** [comment]
Add an explainer

**Comment by mmkal** [comment]
🤖 Added: [`explainers/mobile-builds-and-ota.html`](https://github.com/iterate/iterate/blob/mobile-build-state/explainers/mobile-builds-and-ota.html) — timeline format, four scenarios (JS-only PR, native-change PR, your field-test trap in red, same taps after the fix), per-event system-state YAML on the right, glossary + two appendices (file map; why `Updates.channel` lies). Also opened it in your browser locally. If you wanted it local-only rather than committed, say so and I'll move it to `explainers.ignoreme/`.

**Comment by mmkal** [comment]
🤖 Now also live for phones: https://iterate.iterate.app/explainers/mobile-builds-and-ota (committed to the iterate project's config repo; pinned-to-this-branch variant: https://iterate.iterate.app/explainers/mobile-builds-and-ota?sha=mobile-build-state)

File: apps/mobile/src/lib/build-state-core.ts [hunk 1]
```diff
@@ -0,0 +1,234 @@
+// The app's one answer to "which build am I on, and is it current?" — pure
+// half. Every rule lives here (what channel am I really on, is this binary
+// worth watching for updates, what does an update check MEAN), so the node
+// vitest lane covers all of it without a device. The Expo bindings and the
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state-core.ts]: 🤖 Fixed in 005bf2b86 — "the node vitest lane" → "plain node vitest tests".

File: apps/mobile/src/lib/build-state.ts [hunk 2]
```diff
@@ -0,0 +1,330 @@
+// The app's one answer to "which build am I on, and is it current?" — the
+// Expo/react-query half. Every screen that used to gather these facts itself
+// (Build info, the QR confirm screen, the root layout, the sign-in screen)
+// now reads `useBuildState()` and calls `useBuildActions()`.
+//
+// Absorbs the old build-info.ts, preview-channel.ts and native-install-guard.ts.
+// The rules live next door in build-state-core.ts, which is pure so the node
+// vitest lane covers them without a device.
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts]: 🤖 Fixed in 005bf2b86 — same reword.

File: apps/mobile/src/lib/build-state-core.ts [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+/**
+ * The stamp CI writes into app config `extra.buildInfo`, read off an update
+ * manifest we have NOT downloaded — that is what turns "an update exists" into
+ * "an update with this commit message exists". Shape-checked rather than
+ * trusted: the manifest is remote data.
+ */
+export function stampFromManifest(manifest: unknown): Partial<BuildStamp> {
+  const extra = (manifest as any)?.extra?.expoClient?.extra?.buildInfo;
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state-core.ts:230]: 🤖 Fixed in 005bf2b86 — the cast now carries its justification: remote data with no trustworthy static type; the cast only reaches in, and every field is runtime-checked before use.

File: apps/mobile/src/lib/build-state.ts [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+ * can't OTA. */
+const canOta = () => Updates.isEnabled && !__DEV__;
+
+async function checkForUpdate(): Promise<UpdateCheck> {
+  const result = await Updates.checkForUpdateAsync();
+  if (!result.isAvailable) return { kind: "current" };
+  return {
+    kind: "available",
+    stamp: stampFromManifest(result.manifest),
+    publishedAt: (result.manifest as { createdAt?: string }).createdAt || null,
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts:209]: 🤖 Fixed in 005bf2b86 — explained: the Manifest union includes the embedded shape without createdAt, but an available result always came from the update server, whose manifests carry it; absence reads as unknown publish time.

File: apps/mobile/src/lib/build-state.ts [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+  // to the cache), and hiding it would make the manual button look dead.
+  return {
+    ...withoutCheck(
+      check.isFetching
+        ? { kind: "checking" }
+        : check.isError
+          ? { kind: "error", message: String(check.error) }
+          : check.data || { kind: "idle" },
+    ),
+    ready: override.isSuccess,
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts:292]: 🤖 Real — fixed in 8d8c5a35a. `armed` now waits for the baked-channel read to settle too (a resolved-null still counts as watched — the just-installed boot), and the banner additionally requires `watched`, so a manual Build info check on a main phone can't summon it either.

File: apps/mobile/src/lib/build-state.ts [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+  if (Platform.OS === "web" || !Updates.isEnabled) return nothing;
+  const id = binaryId();
+  const seen = await AsyncStorage.getItem(LAST_SEEN_INSTALL_KEY);
+  if (seen === id) return nothing;
+  await AsyncStorage.setItem(LAST_SEEN_INSTALL_KEY, id);
+  if (seen === null) return nothing;
+  const override = await getChannelOverride();
+  if (override) await setChannelOverride(null);
+  return { binaryChanged: true, clearedOverride: override };
+}
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts:155]: 🤖 Real — fixed in 8d8c5a35a. `setChannelOverride` now invalidates the cached read for every caller, the guard's clear included.

File: apps/mobile/src/lib/build-state.ts [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+  const queryClient = useQueryClient();
+  const switchChannel = useMutation({
+    mutationFn: async ({ channel, revertOnNoUpdate }: SwitchChannelInput) => {
+      const previous = await getChannelOverride();
+      await setChannelOverride(channel);
+      const result = await fetchLatestUpdateAndReload();
+      if (result === "up-to-date" && revertOnNoUpdate) {
+        await setChannelOverride(previous);
+      }
+      return result === "up-to-date" ? ("no-update" as const) : result;
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts]: 🤖 Real, and the nastiest of the four — the stranded-phone trap via the error path. Fixed in 8d8c5a35a: a thrown check/fetch restores the previous override before rethrowing.

File: apps/mobile/src/lib/build-state.ts [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+    retry: false,
+  });
+
+  // No `armed` guard here: an unwatched build's check data can still arrive
+  // via checkNow's imperative fetch (the disabled observer stays subscribed
+  // to the cache), and hiding it would make the manual button look dead.
+  return {
+    ...withoutCheck(
+      check.isFetching
+        ? { kind: "checking" }
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts]: 🤖 Real — fixed in 8d8c5a35a: existing check data now stands during refetches, so the banner survives foregrounds; "checking" only shows before a first verdict.

File: apps/mobile/src/lib/build-state.ts [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+  Updates.setUpdateRequestHeadersOverride(channel ? { "expo-channel-name": channel } : null);
+  if (channel) {
+    await AsyncStorage.setItem(OVERRIDE_KEY, channel);
+  } else {
+    await AsyncStorage.removeItem(OVERRIDE_KEY);
+  }
+  // Every write keeps the cached read honest — including the install guard's
+  // clear, which runs outside any mutation while the banner's read may
+  // already have landed with the old value.
+  void queryClient.invalidateQueries({ queryKey: overrideKey });
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts]: 🤖 Real, and my own doing — the previous fix's invalidate-on-every-write created exactly this. Fixed in 85b1a5106: invalidation moved to the two effectful boundaries (the switch mutation's onSettled — success or error-revert alike — and the install guard's clear). Mid-mutation the cached channel now stays at its pre-switch value, so freshness/auto-continue can't arm early; and the check cache is only dropped on success, since on error the restored override means the old verdict still describes the live channel.

File: apps/mobile/src/lib/build-state.ts [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+      }
+      if (result === "up-to-date" && revertOnNoUpdate) {
+        await setChannelOverride(previous);
+      }
+      return result === "up-to-date" ? ("no-update" as const) : result;
+    },
+    // Settled, not success: the error path REVERTED the override, and the
+    // cache must reflect that too.
+    onSettled: (_data, error) => {
+      void invalidateChannelOverride();
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts:380]: 🤖 Real — fixed in 79a93eee2: the mutation now returns the value it left in place and onSettled writes it into the cache synchronously (then still invalidates to reconcile with storage, which matters on the error path where the restored value isn't in scope).

File: apps/mobile/src/lib/build-state.ts [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+      // The cached check verdict described the OLD channel; drop it rather
+      // than let Build info report the wrong channel's freshness. (Remove,
+      // not invalidate: on an unwatched build nothing would refetch, and
+      // stale-but-displayed is exactly the bug.) Not on error — the override
+      // was restored, so the old verdict still describes the live channel.
+      if (!error) queryClient.removeQueries({ queryKey: checkKey });
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts:386]: 🤖 Real — fixed in 79a93eee2: all callers now share one in-flight checkForUpdateAsync (module-level single-flight), which also covers the guard-boot/QR-scan overlap. The state-machine rejection can't be provoked from this module any more.

File: apps/mobile/src/lib/build-state.ts [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+// call instead.
+let checkInFlight: ReturnType<typeof Updates.checkForUpdateAsync> | null = null;
+function sharedCheckForUpdateAsync() {
+  if (!checkInFlight) {
+    checkInFlight = Updates.checkForUpdateAsync().finally(() => {
+      checkInFlight = null;
+    });
+  }
+  return checkInFlight;
+}
```
[the remark this answers is not in the document]
Comment (mmkal) [reply apps/mobile/src/lib/build-state.ts:197]: 🤖 Real — the single-flight was channel-blind. Fixed in 3cc859a67: checks are stamped with a channel generation (bumped on every override write); same-generation callers join, a newer-generation caller queues behind the stale call instead of overlapping it — so the verdict always describes the channel the caller asked about, and the no-concurrent-check rule still holds.
